core/src/test/scala/org/bykn/bosatsu/EvaluationTest.scala

```diff
@@ -366,4 +366,30 @@ main = [1, 2]
 """), "Foo",
   VList.Cons(VInt(1), VList.Cons(VInt(2), VList.VNil)))
   }
+
+  test("forbid the y-combinator") {
+    evalFail(
+      List("""
+package Y
+
+struct W(fn: W[a, b] -> a -> b)
+
+def call(w0, w1):
+  match w0:
+    W(fn): trace("fn(w1)", fn(w1))
+
+def y(f):
+  g = \w -> \a -> trace("calling f", f(call(w, w), a))
```

Comment:
this does not actually run because we are not lazy enough.

I made some changes to evaluation to try to be lazier, but it didn't work. That's probably a future work: replace the use of cats.Eval in Evaluation with something that has clearer semantics. We likely either want strict or memoizing lazy, it is hard to reason about composing cats.Evals when the whole dag may be repeatedly recomputed.